File: 2012/ CCC '12 S4 - A Coin Game.cpp
// 15/15
// direct translation of my brute force BFS python code
// check the python file for improved bidirectional BFS solution

#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>
#include <string>

using namespace std;

int main() {
    int n;
    while (true) {
        cin >> n;
        if (n == 0) break;

        vector<vector<int>> coins(n);
        for (int i = 0; i < n; i++) {
            int coin;
            cin >> coin;
            coins[i].push_back(coin);
        }

        vector<vector<int>> end(n);
        for (int i = 0; i < n; i++) {
            end[i].push_back(i + 1);
        }

        auto store = [](const vector<vector<int>>& coins) {
            string s;
            for (const auto& stack : coins) {
                for (int coin : stack) {
                    s += to_string(coin);
                }
                s += ' ';
            }
            return s;
        };

        queue<pair<vector<vector<int>>, int>> q;
        q.push({coins, 0});
        unordered_set<string> used;
        used.insert(store(coins));

        bool found = false;
        while (!q.empty()) {
            auto [config, steps] = q.front();
            q.pop();

            if (config == end) {
                cout << steps << endl;
                found = true;
                break;
            }

            for (int i = 0; i < n; i++) {
                if (config[i].empty()) continue;

                if (i != 0 && (config[i - 1].empty() || config[i].back() < config[i - 1].back())) {
                    auto new_config = config;
                    int to_move = new_config[i].back();
                    new_config[i].pop_back();
                    new_config[i - 1].push_back(to_move);

                    string temp = store(new_config);
                    if (used.find(temp) == used.end()) {
                        q.push({new_config, steps + 1});
                        used.insert(temp);
                    }
                }

                if (i != n - 1 && (config[i + 1].empty() || config[i].back() < config[i + 1].back())) {
                    auto new_config = config;
                    int to_move = new_config[i].back();
                    new_config[i].pop_back();
                    new_config[i + 1].push_back(to_move);

                    string temp = store(new_config);
                    if (used.find(temp) == used.end()) {
                        q.push({new_config, steps + 1});
                        used.insert(temp);
                    }
                }
            }
        }

        if (!found) cout << "IMPOSSIBLE" << endl;
    }

    return 0;
}

File: 2012/CCC '12 S4 - A Coin Game.py
# brute force through all combinations
# passes the test cases for junior division
# TLE on senior division, check the c++ version of this code for a solution that passes

from collections import deque

# for each test case
while True:
    n = int(input())
    if n == 0:
        break

    coins = list(map(int, input().split()))
    coins = [[i] for i in coins]  # each element is a stack of coins
    end = [[i] for i in range(1, n+1)]  # the state we are trying to reach
    store = lambda coins: "".join(map(str, coins))  # save space when storing in used

    q = deque()
    q.append((coins, 0))
    used = set()
    used.add(store(coins))

    found = False
    while q:
        config, steps = q.popleft()
        # print(config)

        if config == end:  # is sorted
            print(steps)
            found = True

        for i in range(n):
            if not config[i]:  # nothing to move from here
                continue

            if i != 0 and (not config[i - 1] or config[i][-1] < config[i - 1][-1]):  # try moving left
                new_config = [i.copy() for i in config]
                to_move = new_config[i].pop()
                new_config[i - 1].append(to_move)  # move coin to the left

                temp = store(new_config)
                if temp not in used:
                    q.append((new_config, steps + 1))
                    used.add(temp)

            if i != n - 1 and (not config[i + 1] or config[i][-1] < config[i + 1][-1]):  # try moving right
                new_config = [i.copy() for i in config]
                to_move = new_config[i].pop()
                new_config[i + 1].append(to_move)  # move coin to the right

                temp = store(new_config)
                if temp not in used:
                    q.append((new_config, steps + 1))
                    used.add(temp)

    if not found:
        print("IMPOSSIBLE")
